### PR TITLE
fix: out of order headings display in toc

### LINKED ISSUE #14574: [4.0.0] Table of contents does not work when first cell is of a lower heading priority than succeeding cells
## Description

If the first cell of a notebook is not of the highest level, any cells later in the notebook which are will be excluded from the table of contents. 

## Reproduce

1. New Notebook
2. First cell: markdown, ## heading 1
3. Second cell: markdown, # heading 2
4. Table of contents will render heading 1 but not heading 2

## Expected behavior

All heading-tagged markdown cells should appear in the table of contents, even if they do not have direct parents

## Context

- Operating System and version: Windows 10
- Browser and version:  Firefox 113
- JupyterLab version: 4.0.0

<details><summary>Troubleshoot Output</summary>
<pre>
Paste the output from running `jupyter troubleshoot` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
Paste the output from your browser Javascript console here, if applicable.
22:27:08.262 Unsatisfied version 2.1.1 of shared singleton module @lumino/widgets (required ^1.19.0) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.264 Unsatisfied version 2.0.0 of shared singleton module @lumino/messaging (required ^1.4.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.265 Unsatisfied version 2.1.1 of shared singleton module @lumino/coreutils (required ^1.5.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.267 Unsatisfied version 2.1.1 from @jupyterlab/application-top of shared singleton module @lumino/widgets (required ^1.33.0) [remoteEntry.5cbfa1e4260a81a3d54b.js:1:5956](http://localhost:8888/lab/extensions/jupyterlab-plotly/static/remoteEntry.5cbfa1e4260a81a3d54b.js)
22:27:08.268 Unsatisfied version 2.1.1 of shared singleton module @lumino/coreutils (required ^1.5.3) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.269 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/notebook (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.269 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/codeeditor (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.269 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/settingregistry (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 2.0.0 of shared singleton module @lumino/messaging (required ^1.4.3) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 6.0.0 of shared singleton module @jupyterlab/coreutils (required ^5.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 7.0.0 of shared singleton module @jupyterlab/services (required ^6.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 2.1.1 of shared singleton module @lumino/signaling (required ^1.4.3) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/codemirror (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.270 Unsatisfied version 2.1.1 of shared singleton module @lumino/widgets (required ^1.16.1) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.271 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/apputils (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.271 No satisfying version (^3.0.0) of shared module @jupyterlab/nbformat found in shared scope default.
Available versions: 4.0.0 from @jupyterlab/application-top [remoteEntry.83c796749c1a515caea8.js:1:5228](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.271 Unsatisfied version 2.1.1 of shared singleton module @lumino/disposable (required ^1.4.3) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.272 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/rendermime (required ^3.0.0) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.272 Unsatisfied version 2.0.0 of shared singleton module @lumino/algorithm (required ^1.3.3) [remoteEntry.83c796749c1a515caea8.js:1:4840](http://localhost:8888/lab/extensions/nbdime-jupyterlab/static/remoteEntry.83c796749c1a515caea8.js)
22:27:08.273 Unsatisfied version 2.1.1 of shared singleton module @lumino/signaling (required ^1.4.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.273 Unsatisfied version 2.0.0 of shared singleton module @lumino/algorithm (required ^1.3.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.274 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/logconsole (required ^3.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.274 Unsatisfied version 7.0.0 of shared singleton module @jupyterlab/services (required ^6.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.275 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/notebook (required ^3.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.275 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/settingregistry (required ^3.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.315 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/rendermime (required ^3.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.315 No satisfying version (^3.1.10) of shared module @jupyterlab/outputarea found in shared scope default.
Available versions: 4.0.0 from @jupyterlab/application-top [remoteEntry.f6574b55a223d23e7a04.js:1:6620](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.316 Unsatisfied version 2.0.0 of shared singleton module @lumino/properties (required ^1.2.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.316 Unsatisfied version 2.1.1 of shared singleton module @lumino/disposable (required ^1.4.3) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.316 Unsatisfied version 4.0.0 of shared singleton module @jupyterlab/mainmenu (required ^3.1.10) [remoteEntry.f6574b55a223d23e7a04.js:1:6232](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/remoteEntry.f6574b55a223d23e7a04.js)
22:27:08.672 Failed to create module: package: nbdime-jupyterlab; module: ./extension [index.out.js:16:12](webpack://jupyterlab/application-top/build/index.out.js)
22:27:08.673
Error: Shared module codemirror doesn't exist in shared scope default
    i remoteEntry.83c796749c1a515caea8.js:1
    b remoteEntry.83c796749c1a515caea8.js:1
    m remoteEntry.83c796749c1a515caea8.js:1
    8681 remoteEntry.83c796749c1a515caea8.js:1
    consumes remoteEntry.83c796749c1a515caea8.js:1
    consumes remoteEntry.83c796749c1a515caea8.js:1
    e remoteEntry.83c796749c1a515caea8.js:1
    e remoteEntry.83c796749c1a515caea8.js:1
    extension remoteEntry.83c796749c1a515caea8.js:1
    o remoteEntry.83c796749c1a515caea8.js:1
    F index.out.js:13
    K index.out.js:78
    K index.out.js:75
[index.out.js:585:14](webpack://jupyterlab/application-top/build/index.out.js)
22:27:17.751 Language pack 'English_United States' not valid! [manager.js:49:20](webpack://jupyterlab/application-top/node_modules/@jupyterlab/translation/lib/manager.js)
22:27:17.866 Starting application in workspace: "default" [index.js:397:16](webpack://jupyterlab/application-top/node_modules/@jupyterlab/application-extension/lib/index.js)
22:27:36.222 <source> element has no “src” attribute. Media resource load failed. [lab](http://localhost:8888/lab)
22:27:36.418 All candidate resources failed to load. Media load paused. [lab](http://localhost:8888/lab)
22:27:37.274
Some cookies are misusing the recommended “SameSite“ attribute 87
22:27:39.834 Starting WebSocket: ws://localhost:8888/api/kernels/dbb8c1b2-3941-41fa-9626-9d0dd42b7168 3 [default.js:70:20](webpack://jupyterlab/application-top/node_modules/@jupyterlab/services/lib/kernel/default.js)
22:27:51.827 Uncaught (in promise) TypeError: e.metadata.get is not a function
    _loadFromNotebook http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js:1
    restoreWidgets http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js:1
[523.cf94b68f6edde7eefd61.js:1:5541](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js)
22:43:13.500 Starting WebSocket: ws://localhost:8888/api/kernels/05982139-1d68-4cb6-be7f-f29e4e8b2eea [default.js:70:20](webpack://jupyterlab/application-top/node_modules/@jupyterlab/services/lib/kernel/default.js)
22:43:18.528
Error rendering buttons of the cell toolbar:  TypeError: t.layout is null
    _addToolbar celltoolbartracker.js:124
[celltoolbartracker.js:133:24](webpack://jupyterlab/application-top/node_modules/@jupyterlab/cell-toolbar/lib/celltoolbartracker.js)
22:43:20.568 Uncaught (in promise) TypeError: e.metadata.get is not a function
    _loadFromNotebook http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js:1
    restoreWidgets http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js:1
[523.cf94b68f6edde7eefd61.js:1:5541](http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/523.cf94b68f6edde7eefd61.js)
22:43:29.034
Error rendering buttons of the cell toolbar:  TypeError: t.layout is null
    _addToolbar celltoolbartracker.js:124
[celltoolbartracker.js:133:24](webpack://jupyterlab/application-top/node_modules/@jupyterlab/cell-toolbar/lib/celltoolbartracker.js)
22:43:43.587
Error rendering buttons of the cell toolbar:  TypeError: t.layout is null
    _addToolbar celltoolbartracker.js:124
</pre>
</details>

### PR DESCRIPTION
**Pull request by skyetim**
## References

Fix #14574 

## Code changes

The current codes assume only one tree. The updated code fixes the DFS in `TableOfContentsTree.buildTree` to support multiple trees (out of order headings). 

## User-facing changes

Before
image

After
image

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by fcollonval** [comment]
bot please update galata snapshots

File: packages/toc/src/toctree.tsx [hunk 1]
```diff
@@ -58,41 +58,45 @@ export class TableOfContentsTree extends React.PureComponent<ITableOfContentsTre
...
+        <TableOfContentsItem
+          key={`${current.level}-${currentIndex}-${current.text}`}
+          isActive={
+            !!this.props.activeHeading && current === this.props.activeHeading
+          }
+          heading={current}
+          onMouseDown={this.props.setActiveHeading}
+          onCollapse={this.props.onCollapseChange}
+        >
+          {children}
```
Comment (fcollonval) [inline packages/toc/src/toctree.tsx:86]: Ok there is a bug for leaf element that should be fixed by:

```suggestion
          {children.length ? children : null}
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks for updating the test @skyetim 

So there is actually a small bug; the collapse triangle is shown for all headings even those without children. I made a suggestion to fix it.

**Comment by skyetim** [comment]
bot please update galata snapshots

**Comment by krassowski** [comment]
Closing and re-opening to trigger CI build (stalled after bot update).

File: galata/test/jupyterlab/notebook-run.test.ts-snapshots/notebook-panel-1-jupyterlab-linux.png
Comment (krassowski) [inline galata/test/jupyterlab/notebook-run.test.ts-snapshots/notebook-panel-1-jupyterlab-linux.png:1]: This needs to be reverted.

File: galata/test/jupyterlab/toc-running.test.ts-snapshots/toc-running-indicators-jupyterlab-linux.png
Comment (skyetim) [inline galata/test/jupyterlab/toc-running.test.ts-snapshots/toc-running-indicators-jupyterlab-linux.png:1]: I guess this needs to be removed too?

Comment (krassowski) [reply galata/test/jupyterlab/toc-running.test.ts-snapshots/toc-running-indicators-jupyterlab-linux.png:1]: Either way is fine, it seems that visual regression tests pass now except for one known issue.

**Review by krassowski — APPROVED** [review]
I believe this one is good to go but will wait with merging in case @fcollonval wants to have another look.

**Review by fcollonval — APPROVED** [review]
Thanks @skyetim

**Comment by fcollonval** [comment]
@meeseeksdev please backport to 4.0.x

**Pull request merged**